sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala

```diff
@@ -527,11 +527,22 @@ private[parquet] class ParquetFilters(
         }
 
       case sources.Or(lhs, rhs) =>
+        // The Or predicate is convertible when both of its children can be pushed down.
+        // That is to say, if one/both of the children can be partially pushed down, the Or
+        // predicate can be partially pushed down as well.
+        //
+        // Here is an example used to explain the reason.
+        // Let's say we have
+        // (a1 AND a2) OR (b1 AND b2),
+        // a1 and b1 is convertible, while a2 and b2 is not.
+        // The predicate can be converted as
+        // (a1 OR b1) AND (a1 OR b2) AND (a2 OR b1) AND (a2 OR b2)
```

Comment:
The rule is correct, but it looks a little dangerous to set `true` blindly without considering the given parameter `canPartialPushDownConjuncts`. Could you add a test case for that complex example like `OR` under `NOT` in the deep predicate tree?

Also, it would be great to add more higher level test case in `SQLQuerySuite.scala` to show the benefit of this additional predicate pushdown `a1 OR b1`. Could you add that, too?

Reply:
@dongjoon-hyun nice catch. However, here the `Not` predicate won't have a child as `Or` or `And` predicate because of `BooleanSimplification` optimization rule.
I have created a PR for pushing down `Not` operator before(for double insurance), but seems the PR made thing too complex: https://github.com/apache/spark/pull/22687

> Also, it would be great to add more higher level test case in SQLQuerySuite.scala to show the benefit of this additional predicate pushdown a1 OR b1. Could you add that, too?

How can we verify the predicate is pushed down? Match the `OrcScan` and check the `pushedFilters`? Only Orc V2 can be checked in this way currently.

Reply:
The conversion from `(a1 AND a2) OR (b1 AND b2)` to`(a1 OR b1) AND (a1 OR b2) AND (a2 OR b1) AND (a2 OR b2)` always works no matter it's a top level boolean expression or not.

Reply:
@gengliangwang . You can use  `explain` with `Console.withOut`. What we need is to check `PushedFilters: ` in the plan.

> How can we verify the predicate is pushed down? Match the OrcScan and check the pushedFilters? Only Orc V2 can be checked in this way currently.

@cloud-fan and @gengliangwang . Yes. Of course, I already agreed that the rule is correct. I want to have a test case in a complete end-to-end query form which shows the new benefit clearly. Could you please add a real use case you met?

Reply:
+1 for an end-to-end test